app/src/main/java/com/android/sample/model/user/UserViewModel.kt

```diff
@@ -192,54 +196,94 @@ class UserViewModel(
+  private fun addIngredientToUserFridge(
+      ingredient: Ingredient,
+      quantity: String,
+      expirationDate: LocalDate
+  ) {
+    val newFridgeItem = FridgeItem(ingredient.barCode.toString(), quantity, expirationDate)
+    updateList(_fridgeItems, Pair(newFridgeItem, ingredient), add = true)
   }
 
   /**
-   * Removes an ingredient from the user fridge. More precisely, it will decrease the count of said
-   * ingredient in the fridge.
+   * Method that removes an [Ingredient] from the fridge of the user and updates the user in the
+   * database
    *
    * @param ingredient the ingredient to be removed
-   * @param count the number of ingredients to be removed
    * @throws IllegalArgumentException if the count is greater than the number of ingredients in the
    *   fridge
    */
-  fun removeIngredientFromUserFridge(ingredient: Ingredient, count: Int = 1) {
+  fun removeIngredientFromUserFridge(ingredient: Ingredient) {
     try {
-      val changedElement: Pair<Ingredient, Int> =
-          _fridge.value.first { it.first.barCode == ingredient.barCode }
-
-      if (changedElement.second - count > 0) {
-        updateList(_fridge, changedElement, add = false)
-        updateList(_fridge, Pair(changedElement.first, changedElement.second - count), add = true)
-      } else if (changedElement.second - count == 0) {
-        updateList(_fridge, changedElement, add = false)
-        updateList(_fridge, changedElement, add = false)
-      } else {
-        throw IllegalArgumentException(REMOVED_TOO_MANY_INGREDIENTS_ERROR)
-      }
+      val deletedIngredient: Pair<FridgeItem, Ingredient> =
+          _fridgeItems.value.first { it.first.id == ingredient.barCode.toString() }
+      updateList(_fridgeItems, deletedIngredient, add = false)
     } catch (e: NoSuchElementException) {
       throw IllegalArgumentException(REMOVED_INGREDIENT_NOT_IN_FRIDGE_ERROR)
     }
     updateCurrentUser()
   }
 
+  /**
+   * Method that either add an ingredient to the user's fridge if it does not exist, and updates it
+   * with the new quantity if it is already in the fridge
+   *
+   * @param ingredient the ingredient to be updated/added
+   * @param quantity a valid quantity for the ingredient
+   * @param expirationDate the expiration date of the ingredient
+   * @param scannedItem boolean that the determines if we update via a scanned item or not
+   */
+  fun updateIngredientFromFridge(
+      ingredient: Ingredient,
+      quantity: String,
+      expirationDate: LocalDate,
+      scannedItem: Boolean
+  ) {
+    try {
+      val changedIngredient: Pair<FridgeItem, Ingredient> =
+          _fridgeItems.value.first {
+            (it.first.id == ingredient.barCode.toString()) &&
+                (it.first.expirationDate == expirationDate)
+          }
+      if (scannedItem) { // case there we scan an item that we want to add to the fridge
+        val actualQuantity = regexForQuantity.find(changedIngredient.first.quantity)!!.value
```

Comment:
The line `val actualQuantity = regexForQuantity.find(changedIngredient.first.quantity)!!.value` could potentially throw a `NullPointerException` that isn’t caught by the current `try-catch` block, which could result in the app crashing. 

Additionally, I’m not entirely comfortable with the way this method is implemented. For instance, the use of `_fridgeItems.value.first` can lead to a `NoSuchElementException` if no element is found, as stated in the [documentation](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/first.html). Instead, using the [find](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/find.html) method would avoid throwing an exception if no element is found, as it would simply return `null`. 

This approach would also allow us to eliminate the `try-catch` block entirely by replacing it with a straightforward null check using an `if`. I believe this would make the method cleaner and more maintainable. What are your thoughts on this?

Reply:
Concerning the `find` method of the regex, it is no longer used. Thus, there is no more concerns about a `NullPointerException`.
About the `_fridgeItems.value.first`, I replaced the `.first` with `.find` like you suggested and replaced the `try-catch` block by an `if-else` statement. Tell me what you think about it.

Reply:
LGTM